### ISSUE TITLE
Cannot connect to Rancher when server EC2 instance is re-created

### ISSUE THREAD

**Issue by PeterMaco** [rancher/terraform-k3s-aws-cluster#4]
Hello

I use this TF module to create K3s cluster for Rancher. It creates auto-scaling groups in AWS. When I stop Agent EC2 instances then the auto-scaling group will terminate them and create new instances. This works fine, the Rancher cluster is still functional and I can connect to Rancher Web GUI. But when I stop Server instances, wait for the auto-scaling group to re-create it, then I cannot connect to Rancher anymore. Then I need to destroy everything and re-create whole cluster again. I use AWS Postgress backend for K3s.
I also tried to create 2 instances of Server nodes and it also doesn't work, only one node is registered in Cluster and the other one is shown as unhealthy in AWS load balancer.

Module uses quite old K3s version (0.9.1). Even if it is possible to specify K3s version, I think it will not work with versions 1.X.X as many configuration options were renamed. Are there any plans to support it?

Thanks
Peter

**Comment by drpebcak** [comment]
I wrote this when I worked for rancher, but since then I haven't had access to update it, and nobody else seems to have taken it over.

I have an alternate version here: https://github.com/drpebcak/terraform-k3s-rancher that aims to accomplish a lot of the same goals. I use that version where I work now, so it may not be as broad as this one, but I would welcome PRs and issues there.

As far as multiple k3s server instances go, neither module is written with support of that in mind. The one that lives in my account definitely persists state across recreation of the server ec2 instance though. I use it with RDS, and frequently delete the server instance to let the ASG create a new one.

**Comment by PeterMaco** [comment]
Yes, I know about this alternative version and I tried it yesterday. Unfortunately, I was not able to make the cluster functional because of the invalid certificate, I think I reached the limit on Letsencrypt, so I will have to wait till the next week.

But what I noticed was that default-http-backend didn't want to start because of node selector requires node-role.kubernetes.io/worker label but nodes created by a new version of k3s do not have this label. It looks like some rancher/k3s compatibility issues. 

Regarding multiple k3s instances - why it is not working? Rancher HA document explicitly says that there should be at least 2 instances (https://rancher.com/docs/k3s/latest/en/installation/ha/). Does it require some additional configuration or where the problem is?
I also have configured persistence in RDS but it doesn't work anyway. 

For me, it is quite difficult to investigate any issue there as rancher nodes run in private subnet and I'm not able to see any logs. I noticed that you have an example there which creates bastion box, but I'm not sure how to access it. Example contains ed25519 ssh key, but as I know AWS supports only RSA and I also do not see any port forwarding configuration in bastion box.

**Comment by drpebcak** [comment]
You don't really need the default-http-backend, but you could add those labels to the workers when you create them if you wanted.

Yes, there is additional configuration required for HA - see here https://rancher.com/docs/k3s/latest/en/installation/ha/#3-configure-the-fixed-registration-address
It's not super hard, but that is an extra step.

The bastion is created on a public subnet, so you could just ssh in there. All you need to know is the public IP... You would want to replace the example key with your own, but it doesn't really matter what AWS supports, that key is just being written to `authorized_keys` inside the instance. At that point it's just down to what ssh supports.

**Comment by PeterMaco** [comment]
I tried it again after Letsencrypt limit reset. Now it works fine even when I stop/recreate server EC2 instance. It looks like new version of k3s solved the issue.
The only thing I had to change was default-http-backend selector because worker nodes a missing worker label. 

But terminated nodes are still shown in rancher nodes as Unavailable and it is not possible to delete them. Is there any way how to remove them? 

Thanks
P.

**Comment by drpebcak** [comment]
You can remove them via kubectl. I think the only way to auto remove them would be to go through all the steps to get a cloud provider working

**Comment by PeterMaco** [comment]
Please, how to upgrade Rancher deployed with this terraform module? When changed rancher_version property nothing happens as it only changes user_data in EC2 template. But as I cannot re-create the Rancher cluster nodes because of this issue it cannot be upgraded.

**Comment by Ticiano-mw** [comment]
had the same problems as you @PeterMaco 
 I think we need to do a major refactor to get this working properly... 
I am working on making it possible to have Agent+Server roles on the same EC2 instance, since you only require 2 instance in this case to be HA.

**Comment by drpebcak** [comment]
The way that it's written, you're intended to recreate the master to actually update rancher. I haven't actually used this repo in a while though (about 9 months) so it may not be working quite right anymore.
